chore: add multi-service testing example

octonato opened the pull request:
This is will be the basis of the full docs on how to run multiple services. I will come back to this when we have the Proxy ready.

octonato commented on samples/java-spring-eventsourced-customer-registry/src/main/java/customer/api/CustomerEntity.java:
```diff
@@ -19,17 +19,23 @@
 
   @PostMapping("/create")
-  public Effect<String> create(@RequestBody Customer customer) {
+  public Effect<Confirm> create(@RequestBody Customer customer) {
```
I change this to return `Confirm` instead of raw `String`. The reason is that when calling through the HTTP endpoint, we get: `\"OK\"`. 

Then, when calling from the customer-registry-action, we get: `\"\"OK\"\"`. Not cool.

efgpinto reviewed (approved):
LGTM. I think the _uber_ dockers are not so uber after all and they're fairly understandable. 👍

aludwiko commented on samples/java-protobuf-eventsourced-customer-registry-action/docker-compose-all.yml:
```diff
@@ -0,0 +1,60 @@
+    extra_hosts:
+      - "host.docker.internal:host-gateway"
+    environment:
+      JAVA_TOOL_OPTIONS: > 
```
">" nice, I didn't know that.

aludwiko commented on samples/java-spring-eventsourced-customer-registry/src/main/java/customer/api/CustomerEntity.java:
```diff
@@ -19,17 +19,23 @@
+  public Effect<Confirm> create(@RequestBody Customer customer) {
     logger.info("Creating {}", customer);
+
+    logger.info("what {}", Confirm.done);
```
leftover?

octonato replied:
At least I didn't push the whole sentence. What that f...?

aludwiko reviewed (dismissed):
LGTM.

octonato commented:
I moved this PR back to draft. 

I rebased it with #1606 and #1608. That's all the diff looks huge now. Please ignore it for now. Diff will be much shorter after we merge this two other PR. 

The strategy that I'm taking now is to run a simple integration test that will start a docker-compose running three containers: 

* the proxy for customer-registry
* customer-registry 
* the proxy for customer-registry-subscriber. 

It works when running by hand proving that the docker-compose file is correct, but not yet when running the integration test.

aludwiko commented on samples/java-spring-eventsourced-customer-registry-subscriber/README.md:
```diff
@@ -13,38 +13,49 @@ Use Maven to build your project:
 
-To start the proxy, run the following command from this directory:
+
+To start the application locally, the `klaix-maven-plugin` is used. Use the following command:
```
```suggestion
To start the application locally, the `kalix-maven-plugin` is used. Use the following command:
```

aludwiko commented on samples/java-spring-eventsourced-customer-registry-subscriber/README.md:
```diff
@@ -13,38 +13,49 @@ Use Maven to build your project:
 
-To start the application locally, the `spring-boot-maven-plugin` is used. Use the following command:
+This call is made on the subscriber service and will be forwarded to the `java-spring-eventsourced-customer-registry`
+service.
```
why is `service` in a new line?

octonato replied:
no intentional

aludwiko commented on samples/java-spring-eventsourced-customer-registry-subscriber/README.md:
````diff
@@ -13,38 +13,49 @@ Use Maven to build your project:
 ```
 
-To start the application locally, the `spring-boot-maven-plugin` is used. Use the following command:
+This call is made on the subscriber service and will be forwarded to the `java-spring-eventsourced-customer-registry`
````
I would expect an updated Readme in that service as well.

aludwiko commented on sdk/spring-boot-starter-test/src/main/scala/kalix/spring/boot/KalixConfigurationTest.scala:
```diff
@@ -37,6 +38,7 @@ class KalixConfigurationTest(applicationContext: ApplicationContext) extends Kal
 
   /** WebClient pointing to the proxy. */
   @Bean
+  @Lazy
```
Maybe a comment why this must be lazy?

aludwiko reviewed (commented):
Some minor comments.

aludwiko reviewed (commented):
Some minor comments.

aludwiko reviewed (commented):
Some minor comments.

octonato commented on devtools-internal/src/main/scala/kalix/devtools/impl/DockerComposeUtils.scala:
I can confirm that the issue in CI is related to Linux. 

I build locally using Linux and the jar content was:

```
Archive:  kalix-devtools-internal_2.12-1.2.1-M4-21-d30d4d52-SNAPSHOT.jar
  inflating: META-INF/MANIFEST.MF
   creating: kalix/
   creating: kalix/devtools/
  inflating: kalix/devtools/BuildInfo$.class
  inflating: kalix/devtools/BuildInfo.class
```

octonato commented on .github/workflows/ci.yml:
```diff
@@ -564,9 +564,7 @@ jobs:
-          # FIXME would be good to enable integration tests but currently failing to resolve host.docker.internal
-          # pre_cmd: 'docker-compose -f ../java-spring-eventsourced-customer-registry/docker-compose.yml up -d; (cd ../java-spring-eventsourced-customer-registry/; mvn spring-boot:run &);sleep 10;'
-          - { sample: java-spring-eventsourced-customer-registry-subscriber, it: false }
+          - { sample: java-spring-eventsourced-customer-registry-subscriber, it: true, pre_cmd: 'mvn -f ../java-spring-eventsourced-customer-registry/pom.xml package docker:build' }
```
For the integration tests of customer-registry-subscriber, we need to first locally publish the image for customer-registry.

octonato commented on .gitignore:
```diff
@@ -46,3 +46,4 @@ dependency-reduced-pom.xml
 *.versionsBackup
 regenerate
 .vscode/settings.json
+target-2.12
```
see bellow for explanation about this

octonato commented on build.sbt:
```diff
@@ -348,41 +349,63 @@ def githubUrl(v: String): String = {
+  This is needed because the devTools artifact is also used by the sbt and mvn plugins and therefore needs a 2.12 version.
+  Note that crossbuilding is not an option, because when the built selects 2.13, it will try to build sbt-kalix with 2.13
+ */
+lazy val devToolsInternal =
+  devToolsCommon(
+    project
+      .in(file("devtools"))
+      .settings(
+        name := "kalix-devtools-internal",
+        scalaVersion := Dependencies.ScalaVersionForTooling,
+        // to avoid overwriting the 2.13 version
+        target := baseDirectory.value / "target-2.12"))
```
Maybe some more insight on this.

We want `devTools` to be compiled to both 2.12 and 2.13 so we can use it in integration tests and not only as a dependency of mvn and sbt plugins. 

If we make it a normal cross-build and compile with `+compile`, sbt will first compile everything for 2.12. Meaning dev-tools and the sbt-kalix plugin. 

Then it compiles for 2.13. In that second run, it will build devtools_2.13, but when it reaches the sbt-kalix plugin, it switches back to 2.12 because the plugin requires 2.12. At that moment, we get a Scala version conflict. 

A solution to this problem is to create two modules and let the sbt plugin only depend on `kalix-devtools-internal`. 

My first attempt was to redefine the `sourceDirectory` in the internal module to point to the `devTools` src. Surprisingly enough that didn't work in CI. On Linux, somehow we end up with an almost empty jar (contained BuildInfo only). 

My second attempt, which proved to work, was to at a specific target folder. Hence the `target-2.12` hack here.

octonato replied:
Note that `devTools` and `devToolsInternal` modules are sharing the same directory. `in(file(devtools))`, but using different `scalaVersion` and the latter has a different `target` folder.

franciscolopezsancho replied:
Would it make sense to name them `devTools213` and `devTools212`?

octonato replied:
I did it at some point, but came to the conclusion that it was easier to follow if we said: `devTools` (for general usage) and `devToolsInternal` to be used internally by our own plugins. 

Note that these are only the name of sbt modules. 

Users will eventually see `kalix-devtools_2.13` and  `kalix-devtools-internal_2.12`. And in that case, it's better to have 'internal' in the main. It kind of signals that this is not for general usage.

octonato commented on build.sbt:
```diff
@@ -431,7 +454,7 @@ lazy val codegenCore =
     .settings(
       Compile / javacOptions ++= Seq("--release", "11"),
       Compile / scalacOptions ++= Seq("-release", "11"),
-      scalaVersion := Dependencies.ScalaVersionForDevTools)
```
I renamed it to `tooling` to avoid confusion. DevTools is now also using Scala 2.13.

octonato commented on sbt-plugin/src/main/scala/kalix/sbt/KalixPlugin.scala:
```diff
@@ -73,11 +74,11 @@ object KalixPlugin extends AutoPlugin {
       run / javaOptions ++=
-        dockerComposeUtils.value.readServicePortMappings
-        ++ Seq(
+        dockerComposeUtils.value.localServicePortMappings ++
```
When we configure the locally run service, we use the service port mappings we read from the docker file, but remove any references to hosts. We only need the ports.

octonato commented on maven-java/kalix-maven-plugin/src/main/scala/kalix/RunAllMojo.scala:
```diff
@@ -21,8 +21,8 @@ class RunAllMojo extends RunParameters with DockerParameters {
-      dockerComposeUtils.readUserFunctionPort,
-      dockerComposeUtils.readServicePortMappings,
+      dockerComposeUtils.userFunctionPort,
+      dockerComposeUtils.localServicePortMappings,
```
When we configure the locally run service, we use the service port mappings we read from the docker file, but remove any references to hosts. We only need the ports.

octonato commented on samples/java-spring-eventsourced-customer-registry-subscriber/docker-compose-integration.yml:
```diff
@@ -0,0 +1,42 @@
+      JAVA_TOOL_OPTIONS: >
+        -Dconfig.resource=dev-mode.conf
+        -Dlogback.configurationFile=logback-dev-mode.xml
+        -Dkalix.dev-mode.service-port-mappings.customer-registry=host.docker.internal:9000
```
Proxy 1.1.9-M2 can use this property to do local resolution when running in dev-mode. 

When I added it, I was in doubt if we should accept settings with only the port. Like in, `.customer-registry=9000` as opposed to `.customer-registry=host.docker.internal:9000`

It will be much simpler if we said: `customer-registry=9000`. Then the proxy would add the `host.docker.internal` and the user function would add `localhost`. 

But I'm not sure if other docker tools (eg: colima, podman) will also have `host.docker.internal` as a default. Docker Desktop does it and that's also how it behaves in Linux. I can imagine that all other tools will simply follow it, but not sure.

efgpinto replied:
I think its wise to accept both host and port. Colima didn't support `host.docker.internal` until very recently I believe.

octonato replied:
Yes, indeed. 

Eventually, we could also say. If the host is omitted, it defaults to `host.docker.internal` otherwise to whatever the user declares. We let it like that for now and be less strict later.

octonato reviewed (commented):
I added a bunch of comments to clarify decisions I made. 

Ready for a new round of reviews. 

I will start follow-up PRs on top of this one to add the same test for Java and Scala protobuf SDKs.

octonato commented on samples/java-spring-eventsourced-customer-registry-subscriber/src/it/java/customer/api/CustomerIntegrationTest.java:
```diff
@@ -1,92 +1,171 @@
+    confMap.put("kalix.system.akka.coordinated-shutdown.exit-jvm", "off");
+    // dev-mode should be false when running integration tests
+    confMap.put("kalix.dev-mode.enabled", false);
+    confMap.put("kalix.user-function-interface", "0.0.0.0");
```
Last but not least, this little guy was a major time consumer.

Not needed on Mac, but required for Linux. The test was 'mysteriously' failing in CI because of this. 🤦

octonato replied:
Yes, by default is localhost and it fails for Linux.

franciscolopezsancho commented on samples/java-spring-eventsourced-customer-registry-subscriber/README.md:
````diff
@@ -13,38 +13,46 @@ Use Maven to build your project:
 
 ```shell
-docker-compose up
+mvn kalix:runAll
````
I'm getting 
```
[ERROR] Could not find goal 'runAll' in plugin io.kalix:kalix-maven-plugin:1.2.0 among available goals deploy, generate, help -> [Help 1]
org.apache.maven.plugin.MojoNotFoundException: Could not find goal 'runAll' in plugin io.kalix:kalix-maven-plugin:1.2.0 among available goals deploy, generate, help
```
Although I can see the `goal` in 1.2.0 already

octonato replied:
You have to build the SDK locally. 'runAll' is not released yet.

octonato replied:
If you want to run it locally, you need to first:

`./publishLocally.sh`

Then you need to publish the `customer-register` docker image to your local docker registry. Watch out with that. If you are using Apple M1/2 processors, it will be very slow and the tests will probably not even pass. Let me know if that's the case and I show you the trick to build a arm64 image for `customer-register`. 
I will later send a PR to improve the docker build for those services.

And then you need to update the subscriber pom.xml to use the new snapshot. After that, you can run: `mvn -Pit verify` in the subscriber project.

franciscolopezsancho replied:
roger that

franciscolopezsancho replied:
Works like a charm. This improves DX a lot. 
Regarding the `customer-register` sample. How do you find to update the README to use this feature as well in there so that in both places one uses `mvn kalix:runAll`? So far `java-spring-eventsourced-customer-registry` stills points to the `docker compose up`.

octonato replied:
We will have a follow up PR that will update all the samples and update the docs to explain `mvn kalix:runAll`. This PR is more about testing for local service discovery in dev-mode. 

The `kalix:runAll` and local service discovery are features added in previous PRs. And we were missing the testing part. 

Next step will be to document `kalix:runAll` and how to do multi-service testing.

franciscolopezsancho reviewed (approved):
From what I can gather LGTM
I left some suggestions

efgpinto commented on samples/java-spring-eventsourced-customer-registry-subscriber/src/it/java/customer/api/CustomerIntegrationTest.java:
```diff
@@ -1,92 +1,171 @@
+
     } catch (WebClientRequestException ex) {
-      Assertions.fail("This test requires an external kalix service to be running on localhost:9000 but was not able to reach it.");
+      throw new RuntimeException("This test requires an external kalix service to be running on localhost:9001 but was not able to reach it.");
```
The customer registry still runs on 9000, no?

```suggestion
      throw new RuntimeException("This test requires an external kalix service to be running on localhost:9000 but was not able to reach it.");
```

efgpinto reviewed (approved):
LGTM with tiny suggestions.. 

Btw, PR might need a new rebase as it seems to be showing some changes that are not related with this (and are already in main).

octonato commented:
@efgpinto, it's strange that there were some other commits. I have done a rebase yesterday to incorporate a change done by Andrzej, but it should not show up as an extra commit because it was already in main. 

It gives me the impression that someone has done a force push in main and therefore the commit hash changed. 

Anyway, I removed it and rebase it with main again.

pvlugter commented:
> It gives me the impression that someone has done a force push in main and therefore the commit hash changed.

`Squash and merge` creates a new squashed commit.

octonato commented:
> > It gives me the impression that someone has done a force push in main and therefore the commit hash changed.
> 
> `Squash and merge` creates a new squashed commit.

Yes sure, but I rebased with main so I should have the squashed commit in my branch. That would make sense if I had rebased on a branch from an open PR or if I had done a cherry-pick of some commits. 

Anyway, I start to doubt it now. I can't explain how my branch had different hashes than the main.
